core/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java

```diff
@@ -0,0 +1,248 @@
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.*;
+import org.elasticsearch.common.collect.ImmutableOpenIntMap;
+import org.elasticsearch.common.collect.ImmutableOpenMap;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.CountDown;
+import org.elasticsearch.gateway.AsyncShardFetch;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards;
+import org.elasticsearch.gateway.TransportNodesListGatewayStartedShards.NodeGatewayStartedShards;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+import static org.elasticsearch.action.admin.indices.shards.IndicesShardStoresRequest.*;
+
+/**
+ * Transport action that reads the cluster state for shards with the requested criteria (see {@link ClusterHealthStatus}) of specific indices
+ * and fetches store information from all the nodes using {@link TransportNodesListGatewayStartedShards}
+ */
+public class TransportIndicesShardStoresAction extends TransportMasterNodeReadAction<IndicesShardStoresRequest, IndicesShardStoresResponse> {
+
+    private final TransportNodesListGatewayStartedShards listShardStoresInfo;
+
+    @Inject
+    public TransportIndicesShardStoresAction(Settings settings, TransportService transportService, ClusterService clusterService, ThreadPool threadPool, ActionFilters actionFilters,
+                                             TransportNodesListGatewayStartedShards listShardStoresInfo) {
+        super(settings, IndicesShardStoresAction.NAME, transportService, clusterService, threadPool, actionFilters, IndicesShardStoresRequest.class);
+        this.listShardStoresInfo = listShardStoresInfo;
+    }
+
+    @Override
+    protected String executor() {
+        return ThreadPool.Names.SAME;
+    }
+
+    @Override
+    protected IndicesShardStoresResponse newResponse() {
+        return new IndicesShardStoresResponse();
+    }
+
+    @Override
+    protected void masterOperation(IndicesShardStoresRequest request, ClusterState state, ActionListener<IndicesShardStoresResponse> listener) {
+        final RoutingTable routingTables = state.routingTable();
+        final RoutingNodes routingNodes = state.routingNodes();
+        final String[] concreteIndices = state.metaData().concreteIndices(request.indicesOptions(), request.indices());
+        final Set<ShardId> shardIdsToFetch = new HashSet<>();
+
+        // collect relevant shard ids of the requested indices for fetching store infos
+        for (String index : concreteIndices) {
+            for (IndexShardRoutingTable routing : routingTables.index(index)) {
```

Comment:
Thanks for the suggestion.
Now, we use `ClusterShardHealth` to figure out the shard health status in `TransportIndicesShardStoresAction`